ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,543 @@
+struct RequestManager {
+    /// Total completed requests
+    total_completed_requests: u64,
+    /// Total pending requests
+    total_pending_requests: u64,
+
+    /// Pending requests are either:
+    /// * Skipped if they are `RequestMetadata::Data` and exist in the database
+    /// * Completed if the database has keys
+    /// * Converted into one network request
+    pending_requests: HashMap<H256, RequestMetadata>,
+    /// Completed requests
+    responses: Vec<(RequestMetadata, Vec<u8>)>,
+}
+
+impl RequestManager {
+    fn add(&mut self, hash: H256, metadata: RequestMetadata) {
+        let old_metadata = self.pending_requests.insert(hash, metadata);
+
+        if let Some(old_metadata) = old_metadata {
+            debug_assert_eq!(metadata, old_metadata);
+        } else {
+            self.total_pending_requests += 1;
+        }
+    }
+
+    async fn request(
+        &mut self,
+        network: &mut NetworkService,
+        db: &Database,
+    ) -> Option<Vec<(RequestMetadata, Vec<u8>)>> {
+        let pending_network_requests = self.handle_pending_requests(db);
+
+        if !pending_network_requests.is_empty() {
+            let request = pending_network_requests.keys().copied().collect();
+            let request_id = network.db_sync().request(db_sync::Request(request));
+
+            let result = loop {
+                let event = network
+                    .next()
+                    .await
+                    .expect("network service stream is infinite");
+
+                if let NetworkEvent::DbResponse {
+                    request_id: rid,
+                    result,
+                } = event
+                {
+                    debug_assert_eq!(rid, request_id, "unknown request id");
+                    break result;
+                }
+            };
+
+            match result {
+                Ok(response) => {
+                    self.handle_response(pending_network_requests, response, db);
+                }
+                Err((request, err)) => {
+                    network.db_sync().retry(request);
+                    self.pending_requests.extend(pending_network_requests);
+                    log::warn!("{request_id:?} failed: {err}. Retrying...");
+                }
+            }
+        }
+
+        let continue_processing = !(self.pending_requests.is_empty() && self.responses.is_empty());
+        if continue_processing {
+            let responses: Vec<_> = self.responses.drain(..).collect();
+            self.total_completed_requests += responses.len() as u64;
+            Some(responses)
+        } else {
+            None
+        }
+    }
+
+    fn handle_pending_requests(&mut self, db: &Database) -> HashMap<H256, RequestMetadata> {
+        let mut pending_requests = HashMap::new();
+        for (hash, metadata) in self.pending_requests.drain() {
+            if metadata.is_data() && db.has_hash(hash) {
+                self.total_completed_requests += 1;
```

Comment:
tbh it seems a bit incorrect for me to anyway add query in manager even if you have requested hash from the very beginning and treat it as completed request